### ISSUE TITLE
fails to use gcc4.8.1 to compile on AIX6.1

### ISSUE THREAD

**Issue by shaolila** [numpy/numpy#5474]
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building py_modules sources
building library "npymath" sources
error: don't know how to compile C/C++ code on platform 'posix' with 'gcc' compiler

**Comment by rgommers** [comment]
Can you put the full build log in a Gist and link to it? And give us the output of ``gcc --version``?

**Comment by rgommers** [comment]
And did you manage to build an older version of numpy on this system, or have you never tried before?

**Comment by shaolila** [comment]
bash-4.2# gcc -v 
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/opt/freeware/libexec/gcc/powerpc-ibm-aix6.1.0.0/4.8.1/lto-wrapper
Target: powerpc-ibm-aix6.1.0.0
Configured with: ../configure --with-as=/usr/bin/as --with-ld=/usr/bin/ld --enable-languages=c,c++,fortran --prefix=/opt/freeware --mandir=/opt/freeware/man --infodir=/opt/freeware/info --enable-threads --enable-version-specific-runtime-libs --enable-decimal-float=dpd --host=powerpc-ibm-aix6.1.0.0 --target=powerpc-ibm-aix6.1.0.0
Thread model: aix
gcc version 4.8.1 (GCC) 
bash-4.2# oslevel
6.1.0.0

**Comment by shaolila** [comment]
@rgommers  yes,i tried numpy1.8.2 instead of 1.9.1 ,they all the same.

**Comment by juliantaylor** [comment]
this error comes from a very basic test of compiler functionality (and the output is useless, we should fix that ...)
it tries to compile a c file with a main in it and that failed, does that work outside of numpy? do yo have some compiler/flag setting environment variables?

**Comment by juliantaylor** [comment]
hm actually this is something else than what I though, which python version are you using?
looks like python distutils does not work for this platform and numpy has no fallback.

**Comment by juliantaylor** [comment]
get_default_compiler should be returning unix not gcc

**Comment by shaolila** [comment]
python version is 2.7.6  , i can use xlc to compile numpy on aix6.1 successfully,but when i export CC=gcc to change the default compiler to gcc ,it comes this problem as above.

**Comment by rgommers** [comment]
I'd still like to see the full build log, and get the exact command used to start the build.

**Comment by shaolila** [comment]
pip install numpy1.9.1.tar  or python setup.py build all comes this problem as above.

**Comment by shaolila** [comment]
numpy-env)bash-4.2# pip install numpy-1.9.1.tar 
Processing ./numpy-1.9.1.tar
    Running from numpy source directory.
Installing collected packages: numpy
  Running setup.py install for numpy
    Running from numpy source directory.
    /bin/sh: svnversion:  not found.
    non-existing path in 'numpy/distutils': 'site.cfg'
    non-existing path in 'numpy/f2py': 'docs'
    non-existing path in 'numpy/f2py': 'f2py.1'
    F2PY Version 2
    blas_opt_info:
    blas_mkl_info:
      libraries mkl,vml,guide not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    openblas_info:
      libraries openblas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    atlas_blas_threads_info:
    Setting PTATLAS=ATLAS
      libraries ptf77blas,ptcblas,atlas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    atlas_blas_info:
      libraries f77blas,cblas,atlas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1526: UserWarning:
        Atlas (http://math-atlas.sourceforge.net/) libraries not found.
        Directories to search for the libraries can be specified in the
        numpy/distutils/site.cfg file (section [atlas]) or by setting
        the ATLAS environment variable.
      warnings.warn(AtlasNotFoundError.__doc__)
    blas_info:
      FOUND:
        libraries = ['blas']
        library_dirs = ['/usr/lib']
        language = f77
      FOUND:
        libraries = ['blas']
        library_dirs = ['/usr/lib']
        define_macros = [('NO_ATLAS_INFO', 1)]
        language = f77
    /bin/sh: svnversion:  not found.
    non-existing path in 'numpy/lib': 'benchmarks'
    lapack_opt_info:
    openblas_lapack_info:
      libraries openblas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    lapack_mkl_info:
    mkl_info:
      libraries mkl,vml,guide not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
      NOT AVAILABLE
    atlas_threads_info:
    Setting PTATLAS=ATLAS
      libraries ptf77blas,ptcblas,atlas not found in /numpy-assemly/numpy-env/lib
      libraries lapack_atlas not found in /numpy-assemly/numpy-env/lib
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib/
      libraries lapack_atlas not found in /usr/lib/
    numpy.distutils.system_info.atlas_threads_info
      NOT AVAILABLE
    atlas_info:
      libraries f77blas,cblas,atlas not found in /numpy-assemly/numpy-env/lib
      libraries lapack_atlas not found in /numpy-assemly/numpy-env/lib
      libraries f77blas,cblas,atlas not found in /usr/lib
      libraries lapack_atlas not found in /usr/lib
      libraries f77blas,cblas,atlas not found in /usr/lib/
      libraries lapack_atlas not found in /usr/lib/
    numpy.distutils.system_info.atlas_info
      NOT AVAILABLE
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1432: UserWarning:
        Atlas (http://math-atlas.sourceforge.net/) libraries not found.
        Directories to search for the libraries can be specified in the
        numpy/distutils/site.cfg file (section [atlas]) or by setting
        the ATLAS environment variable.
      warnings.warn(AtlasNotFoundError.__doc__)
    lapack_info:
      libraries lapack not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
      NOT AVAILABLE
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1443: UserWarning:
        Lapack (http://www.netlib.org/lapack/) libraries not found.
        Directories to search for the libraries can be specified in the
        numpy/distutils/site.cfg file (section [lapack]) or by setting
        the LAPACK environment variable.
      warnings.warn(LapackNotFoundError.__doc__)
    lapack_src_info:
      NOT AVAILABLE
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1446: UserWarning:
        Lapack (http://www.netlib.org/lapack/) sources not found.
        Directories to search for the sources can be specified in the
        numpy/distutils/site.cfg file (section [lapack_src]) or by setting
        the LAPACK_SRC environment variable.
      warnings.warn(LapackSrcNotFoundError.__doc__)
      NOT AVAILABLE
    /opt/freeware/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'define_macros'
      warnings.warn(msg)
    unifing config_cc, config, build_clib, build_ext, build commands --compiler options
    unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
    build_src
    building py_modules sources
    building library "npymath" sources
    error: don't know how to compile C/C++ code on platform 'posix' with 'gcc' compiler
    Complete output from command /numpy-assemly/numpy-env/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-ow5Ndm-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-EAZ0MR-record/install-record.txt --single-version-externally-managed --compile --install-headers /numpy-assemly/numpy-env/include/site/python2.7:
    Running from numpy source directory.
    
    /bin/sh: svnversion:  not found.
    
    non-existing path in 'numpy/distutils': 'site.cfg'
    
    non-existing path in 'numpy/f2py': 'docs'
    
    non-existing path in 'numpy/f2py': 'f2py.1'
    
    F2PY Version 2
    
    blas_opt_info:
    
    blas_mkl_info:
    
      libraries mkl,vml,guide not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    openblas_info:
    
      libraries openblas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    atlas_blas_threads_info:
    
    Setting PTATLAS=ATLAS
    
      libraries ptf77blas,ptcblas,atlas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    atlas_blas_info:
    
      libraries f77blas,cblas,atlas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1526: UserWarning:
    
        Atlas (http://math-atlas.sourceforge.net/) libraries not found.
    
        Directories to search for the libraries can be specified in the
    
        numpy/distutils/site.cfg file (section [atlas]) or by setting
    
        the ATLAS environment variable.
    
      warnings.warn(AtlasNotFoundError.__doc__)
    
    blas_info:
    
      FOUND:
    
        libraries = ['blas']
    
        library_dirs = ['/usr/lib']
    
        language = f77
    
    
    
      FOUND:
    
        libraries = ['blas']
    
        library_dirs = ['/usr/lib']
    
        define_macros = [('NO_ATLAS_INFO', 1)]
    
        language = f77
    
    
    
    /bin/sh: svnversion:  not found.
    
    non-existing path in 'numpy/lib': 'benchmarks'
    
    lapack_opt_info:
    
    openblas_lapack_info:
    
      libraries openblas not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    lapack_mkl_info:
    
    mkl_info:
    
      libraries mkl,vml,guide not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
      NOT AVAILABLE
    
    
    
    atlas_threads_info:
    
    Setting PTATLAS=ATLAS
    
      libraries ptf77blas,ptcblas,atlas not found in /numpy-assemly/numpy-env/lib
    
      libraries lapack_atlas not found in /numpy-assemly/numpy-env/lib
    
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib
    
      libraries lapack_atlas not found in /usr/lib
    
      libraries ptf77blas,ptcblas,atlas not found in /usr/lib/
    
      libraries lapack_atlas not found in /usr/lib/
    
    numpy.distutils.system_info.atlas_threads_info
    
      NOT AVAILABLE
    
    
    
    atlas_info:
    
      libraries f77blas,cblas,atlas not found in /numpy-assemly/numpy-env/lib
    
      libraries lapack_atlas not found in /numpy-assemly/numpy-env/lib
    
      libraries f77blas,cblas,atlas not found in /usr/lib
    
      libraries lapack_atlas not found in /usr/lib
    
      libraries f77blas,cblas,atlas not found in /usr/lib/
    
      libraries lapack_atlas not found in /usr/lib/
    
    numpy.distutils.system_info.atlas_info
    
      NOT AVAILABLE
    
    
    
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1432: UserWarning:
    
        Atlas (http://math-atlas.sourceforge.net/) libraries not found.
    
        Directories to search for the libraries can be specified in the
    
        numpy/distutils/site.cfg file (section [atlas]) or by setting
    
        the ATLAS environment variable.
    
      warnings.warn(AtlasNotFoundError.__doc__)
    
    lapack_info:
    
      libraries lapack not found in ['/numpy-assemly/numpy-env/lib', '/usr/lib', '/usr/lib/']
    
      NOT AVAILABLE
    
    
    
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1443: UserWarning:
    
        Lapack (http://www.netlib.org/lapack/) libraries not found.
    
        Directories to search for the libraries can be specified in the
    
        numpy/distutils/site.cfg file (section [lapack]) or by setting
    
        the LAPACK environment variable.
    
      warnings.warn(LapackNotFoundError.__doc__)
    
    lapack_src_info:
    
      NOT AVAILABLE
    
    
    
    /tmp/pip-ow5Ndm-build/numpy/distutils/system_info.py:1446: UserWarning:
    
        Lapack (http://www.netlib.org/lapack/) sources not found.
    
        Directories to search for the sources can be specified in the
    
        numpy/distutils/site.cfg file (section [lapack_src]) or by setting
    
        the LAPACK_SRC environment variable.
    
      warnings.warn(LapackSrcNotFoundError.__doc__)
    
      NOT AVAILABLE
    
    
    
    /opt/freeware/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'define_macros'
    
      warnings.warn(msg)
    
    running install
    
    running build
    
    running config_cc
    
    unifing config_cc, config, build_clib, build_ext, build commands --compiler options
    
    running config_fc
    
    unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
    
    running build_src
    
    build_src
    
    building py_modules sources
    
    creating build
    
    creating build/src.aix-6.1-2.7
    
    creating build/src.aix-6.1-2.7/numpy
    
    creating build/src.aix-6.1-2.7/numpy/distutils
    
    building library "npymath" sources
    
    error: don't know how to compile C/C++ code on platform 'posix' with 'gcc' compiler
    
    ----------------------------------------
    Command "/numpy-assemly/numpy-env/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-ow5Ndm-build/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-EAZ0MR-record/install-record.txt --single-version-externally-managed --compile --install-headers /numpy-assemly/numpy-env/include/site/python2.7" failed with error code 1 in /tmp/pip-ow5Ndm-build

**Comment by rgommers** [comment]
The error comes from this part of ``ccompiler.py``:

    try:
        if compiler is None:
            compiler = get_default_compiler(plat)
        (module_name, class_name, long_description) = compiler_class[compiler]
    except KeyError:
        msg = "don't know how to compile C/C++ code on platform '%s'" % plat
        if compiler is not None:
            msg = msg + " with '%s' compiler" % compiler
        raise DistutilsPlatformError(msg)

I've tried ``export CC=gcc`` and building numpy; for me the ``compiler`` variable is always ``unix``. ``gcc`` is never in the ``compiler_class`` dict, hence the ``KeyError`` that triggers your issue. It would help if you could figure out why on your system ``gcc`` ends up in the ``compiler`` variable.

**Comment by shaolila** [comment]
thanks @rgommers  thanks @juliantaylor  thanks @certik 
now i run into another problem when use gcc4.8.1 to compile on AIX6.1 ，please help me ...
error: Command "gcc -DNDEBUG -g -O0 -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -fno-tree-dominator-opts -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Inumpy/core/include -Ibuild/src.aix-6.1-2.7/numpy/core/include/numpy -Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/include -I/opt/freeware/include/python2.7 -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -c build/src.aix-6.1-2.7/numpy/core/src/npysort/heapsort.c -o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/npysort/heapsort.o" failed with exit status 1

**Comment by rgommers** [comment]
@shaolila it looks like you figured out what went wrong with the first error. Can you explain, so the next person that runs into the same issue can learn from it?

**Comment by shaolila** [comment]
@rgommers  In fact, I did not solve the first problem ,i just use another aix ,then it comes the new problem.

**Comment by rgommers** [comment]
Ah, ok. Both AIX 6.1 but some small differences?

**Comment by rgommers** [comment]
Here is the relevant part of the build log you sent me:

creating build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/npysort

	compile options: '-Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Inumpy/core/include -Ibuild/src.aix-6.1-2.7/numpy/core/include/numpy -Inumpy/core/src/private -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/include -I/opt/freeware/include/python2.7 -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -Ibuild/src.aix-6.1-2.7/numpy/core/src/private -c'

	gcc: build/src.aix-6.1-2.7/numpy/core/src/npysort/heapsort.c

	In file included from /opt/freeware/include/python2.7/Python.h:33:0,

		         from numpy/core/src/private/npy_sort.h:5,

		         from numpy/core/src/npysort/heapsort.c.src:32:

	/opt/freeware/lib/gcc/powerpc-ibm-aix6.1.0.0/4.8.1/include-fixed/stdio.h:542:12: error: conflicting types for 'fgetpos64'

	 extern int fgetpos64(FILE *, fpos64_t *);

		    ^

	In file included from /opt/freeware/include/python2.7/Python.h:33:0,

		         from numpy/core/src/private/npy_sort.h:5,

		         from numpy/core/src/npysort/heapsort.c.src:32:

	/opt/freeware/lib/gcc/powerpc-ibm-aix6.1.0.0/4.8.1/include-fixed/stdio.h:337:12: note: previous declaration of 'fgetpos64' was here

	 extern int fgetpos(FILE *__restrict__, fpos_t *__restrict__);

		    ^

	In file included from /opt/freeware/include/python2.7/Python.h:33:0,

		         from numpy/core/src/private/npy_sort.h:5,

		         from numpy/core/src/npysort/heapsort.c.src:32:

	/opt/freeware/lib/gcc/powerpc-ibm-aix6.1.0.0/4.8.1/include-fixed/stdio.h:545:12: error: conflicting types for 'fseeko64'

	 extern int fseeko64(FILE *, off64_t, int);

The same bug as https://svn.boost.org/trac/boost/ticket/10122. 

From http://bugs.python.org/issue11184 I can't easily see into what bugfix Python 2.7 release that patch went, but it was a backport. What is the exact Python version you're using here (2.7.??)?

**Comment by shaolila** [comment]
bash-4.2# python -V
Python 2.7.6

**Comment by shaolila** [comment]
@rgommers this solve  the first problem  #python setup.py build  --compiler=unix  , then it comes the second one.

**Comment by rgommers** [comment]
Ah, took me a while to find it because this is fixed in numpy master, in 661c4a5fb. For 1.9.1, ``heapsort.c.src`` starts with:

    #include <stdlib.h>
    #include "npy_sort.h"
    #include "npysort_common.h"

while in master it is:

    #include "npy_sort.h"
    #include "npysort_common.h"
    #include <stdlib.h>

You can either edit ``heapsort.c.src`` by hand to fix this issue, or download and build numpy master. We'd anyway be interested to hear if ``master`` builds fine on AIX or if there are any other issues.

**Comment by rgommers** [comment]
@juliantaylor want to add that commit to your 1.9.2 backport PR?

**Comment by juliantaylor** [comment]
oh this is gh-4723 good its fixed
@shaolila if you can confirm the commit fixes the issue I can put it into 1.9.2

**Comment by shaolila** [comment]
@rgommers  thank you for your advice ,i fixed the second one ,then ,i comes the third one .....

error: Command "/opt/freeware/lib/python2.7/config/ld_so_aix /usr/vac/bin/xlc_r -DAIX_GENUINE_CPLUSCPLUS -D_LINUX_SOURCE_COMPAT -D_LARGE_FILES -q32 -qbitfields=signed -qmaxmem=70000 -qalloca -bmaxdata:0x80000000 -Wl,-brtl -bI:/opt/freeware/lib/python2.7/config/python.exp -L. -L/usr/lib/threads -L/opt/freeware/lib -L/usr/lib build/temp.aix-6.1-2.7/numpy/core/src/multiarray/alloc.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/arrayobject.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/arraytypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_scalar.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_array.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/buffer.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/calculation.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/common.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert_datatype.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/conversion_utils.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ctors.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_strings.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busday.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busdaycal.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/descriptor.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/dtype_transfer.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/einsum.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/flagsobject.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/getset.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/hashdescr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/item_selection.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/iterators.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/lowlevel_strided_loops.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/mapping.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/methods.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/multiarraymodule.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/nditer_templ.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_api.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_constr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_pywrap.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/number.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpymemoryview.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpyos.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/refcount.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/sequence.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/shape.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/scalarapi.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/scalartypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/usertypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ucsnarrow.o -L/opt/freeware/lib -Lbuild/temp.aix-6.1-2.7 -lnpymath -lnpysort -lm -o build/lib.aix-6.1-2.7/numpy/core/multiarray.so" failed with exit status 8

**Comment by rgommers** [comment]
@shaolila is that with numpy master or the local fix? And can you send me the full build log again?

**Comment by shaolila** [comment]
@rgommers   the local fix....ok,the build log  is too long to paste,so i send a email with an attachment,please check it out.

**Comment by pv** [comment]
@shaolila: it's recommended to paste long outputs to
https://gist.github.com and then give the link here

**Comment by juliantaylor** [comment]
executing that failed commandline manually usually gives the best error message

**Comment by rgommers** [comment]
https://gist.github.com/rgommers/a7c7c4cb9353a5197076

    /opt/freeware/lib/python2.7/config/ld_so_aix /usr/vac/bin/xlc_r -DAIX_GENUINE_CPLUSCPLUS -D_LINUX_SOURCE_COMPAT -D_LARGE_FILES -q32 -qbitfields=signed -qmaxmem=70000 -qalloca -bmaxdata:0x80000000 -Wl,-brtl -bI:/opt/freeware/lib/python2.7/config/python.exp -L. -L/usr/lib/threads -L/opt/freeware/lib -L/usr/lib build/temp.aix-6.1-2.7/numpy/core/src/multiarray/alloc.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/arrayobject.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/arraytypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_scalar.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_array.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/buffer.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/calculation.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/common.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert_datatype.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/conversion_utils.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ctors.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_strings.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busday.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busdaycal.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/descriptor.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/dtype_transfer.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/einsum.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/flagsobject.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/getset.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/hashdescr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/item_selection.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/iterators.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/lowlevel_strided_loops.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/mapping.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/methods.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/multiarraymodule.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/nditer_templ.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_api.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_constr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_pywrap.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/number.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpymemoryview.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpyos.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/refcount.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/sequence.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/shape.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/scalarapi.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/scalartypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/usertypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ucsnarrow.o -L/opt/freeware/lib -Lbuild/temp.aix-6.1-2.7 -lnpymath -lnpysort -lm -o build/lib.aix-6.1-2.7/numpy/core/multiarray.so
    ld: 0711-224 WARNING: Duplicate symbol: __itab
    ld: 0711-224 WARNING: Duplicate symbol: zero
    ld: 0711-224 WARNING: Duplicate symbol: tantab
    ld: 0711-224 WARNING: Duplicate symbol: guesses
    ld: 0711-224 WARNING: Duplicate symbol: __tab
    ld: 0711-224 WARNING: Duplicate symbol: signgam
    ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
    ld: 0711-317 ERROR: Undefined symbol: .__moddi3
    ld: 0711-317 ERROR: Undefined symbol: .__divdi3
    ld: 0711-317 ERROR: Undefined symbol: .__umoddi3
    ld: 0711-317 ERROR: Undefined symbol: .__udivdi3
    ld: 0711-317 ERROR: Undefined symbol: .__floatundisf
    ld: 0711-317 ERROR: Undefined symbol: .__cmpdi2
    ld: 0711-317 ERROR: Undefined symbol: .__floatdisf
    ld: 0711-224 WARNING: Duplicate symbol: __itab
    ld: 0711-224 WARNING: Duplicate symbol: zero
    ld: 0711-224 WARNING: Duplicate symbol: tantab
    ld: 0711-224 WARNING: Duplicate symbol: guesses
    ld: 0711-224 WARNING: Duplicate symbol: __tab
    ld: 0711-224 WARNING: Duplicate symbol: signgam
    ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
    ld: 0711-317 ERROR: Undefined symbol: .__moddi3
    ld: 0711-317 ERROR: Undefined symbol: .__divdi3
    ld: 0711-317 ERROR: Undefined symbol: .__umoddi3
    ld: 0711-317 ERROR: Undefined symbol: .__udivdi3
    ld: 0711-317 ERROR: Undefined symbol: .__floatundisf
    ld: 0711-317 ERROR: Undefined symbol: .__cmpdi2
    ld: 0711-317 ERROR: Undefined symbol: .__floatdisf
    error: Command "/opt/freeware/lib/python2.7/config/ld_so_aix /usr/vac/bin/xlc_r -DAIX_GENUINE_CPLUSCPLUS -D_LINUX_SOURCE_COMPAT -D_LARGE_FILES -q32 -qbitfields=signed -qmaxmem=70000 -qalloca -bmaxdata:0x80000000 -Wl,-brtl -bI:/opt/freeware/lib/python2.7/config/python.exp -L. -L/usr/lib/threads -L/opt/freeware/lib -L/usr/lib build/temp.aix-6.1-2.7/numpy/core/src/multiarray/alloc.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/arrayobject.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/arraytypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_scalar.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/array_assign_array.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/buffer.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/calculation.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/common.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/convert_datatype.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/conversion_utils.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ctors.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_strings.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busday.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/datetime_busdaycal.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/descriptor.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/dtype_transfer.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/einsum.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/flagsobject.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/getset.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/hashdescr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/item_selection.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/iterators.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/lowlevel_strided_loops.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/mapping.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/methods.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/multiarraymodule.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/nditer_templ.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_api.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_constr.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/nditer_pywrap.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/number.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpymemoryview.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/numpyos.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/refcount.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/sequence.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/shape.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/scalarapi.o build/temp.aix-6.1-2.7/build/src.aix-6.1-2.7/numpy/core/src/multiarray/scalartypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/usertypes.o build/temp.aix-6.1-2.7/numpy/core/src/multiarray/ucsnarrow.o -L/opt/freeware/lib -Lbuild/temp.aix-6.1-2.7 -lnpymath -lnpysort -lm -o build/lib.aix-6.1-2.7/numpy/core/multiarray.so" failed with exit status 8

**Comment by juliantaylor** [comment]
urg no idea what thats about, looks like someone with knowledge of AIX will have to check that out.

**Comment by rgommers** [comment]
There are several posts that say linking should be done with gcc not ld, so a flag ``-lgcc`` would help. However says that that brings more problems. This seems to be a common issue on AIX with no clear solution.

**Comment by rgommers** [comment]
Do you need to use gcc? xlc seems to work for you.

**Comment by shaolila** [comment]
@rgommers  haha,yes,xlc indeed works,but I must use gcc,because xlc need license....

**Comment by shaolila** [comment]
@rgommers  @juliantaylor  @pv  @certik  @cournape hi guys , Thank you all, I've figured out the third problem with  the  following commands ,it finally use gcc from beginning to end.
---------------------------------------------------------------------------------------------------------------------
# export 'CC=gcc -lgcc'
# export LDSHARED="/opt/freeware/lib/python2.7/config/ld_so_aix gcc -bI:/opt/freeware/lib/python2.7/config/python.exp"
# export CFLAGS="-pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -fno-tree-dominator-opts"
----------------------------------------------------------------------------------------------------------------------
but ,it comes the fourth problem....
https://gist.github.com/shaolila/2f0196b80f459ed84328

**Comment by shaolila** [comment]
@rgommers  @juliantaylor  @pv  @certik  @cournape 

hi guys , Thank you all, I've figured out the third problem with the following commands ,it finally use gcc from beginning to end.
-------------------------------------------------------------------------------------------
export 'CC=gcc -lgcc'

export LDSHARED="/opt/freeware/lib/python2.7/config/ld_so_aix gcc -bI:/opt/freeware/lib/python2.7/config/python.exp"

export CFLAGS="-pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -fno-tree-dominator-opts"
--------------------------------------------------------------------------------------------
but ,it comes the fourth problem....
https://gist.github.com/shaolila/2f0196b80f459ed84328

**Comment by shaolila** [comment]
@rgommers  @juliantaylor  @pv  @certik  @cournape 

hi guys , Thank you all, I've figured out the third problem with the following commands ,it finally use gcc from beginning to end.
--------------------------------------------------------------------------------------------------
export 'CC=gcc -lgcc'

export LDSHARED="/opt/freeware/lib/python2.7/config/ld_so_aix gcc -bI:/opt/freeware/lib/python2.7/config/python.exp"

export CFLAGS="-pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -fno-tree-dominator-opts"
-----------------------------------------------------------------------------------------------------
but ,it comes the fourth problem....
https://gist.github.com/shaolila/2f0196b80f459ed84328

**Comment by rgommers** [comment]
This is the same kind of error as before, also already fixed in master: ee9f85bb91. You can copy that fix. 

@juliantaylor can you add it to gh-5500 again?

**Comment by shaolila** [comment]
when i use numpy-master instead of numpy1.9.1  ,it comes the followings information that also contains errors....
https://gist.github.com/shaolila/262bdc121da7d4992330

**Comment by shaolila** [comment]
And furthermore, when i use numpy1.9.1 after copy the fix in ee9f85b ,it also has errors as master  does.

**Comment by rgommers** [comment]
same issue again. if you get a compile error for a file, check if it includes first ``Python.h`` and then ``math.h`` and ``stdio.h``. Changing to that order will fix things.

Put this fix in a PR: gh-5515.

**Comment by shaolila** [comment]
thanks, i will check all of this kind of error.......and, but , the problems was not just looks like only due to it .
from line 195 to line 201 in https://gist.github.com/shaolila/262bdc121da7d4992330,something about "symbol: .exp" goes wrong.

**Comment by rgommers** [comment]
@shaolila any error in ``_configtest.c`` you can ignore. Those are small test compiles done by the build system to check things before starting the real build.

**Comment by shaolila** [comment]
So that explains it ........
from line 7608 to line 7746 in https://gist.github.com/shaolila/262bdc121da7d4992330 , it seems compilation about fftpack.c went wrong , I already check it  includes first

**Comment by rgommers** [comment]
It seems to go wrong on ``#include <numpy/ndarraytypes.h>``, did you also move that up before ``math.h`` and ``stdio.h``?

**Comment by rgommers** [comment]
No more follow-up, and I think this last error was fixed by https://github.com/numpy/numpy/commit/214bfd432e2f. So closing.

If there's more going wrong on AIX, opening a new issue would be better.
